pkg/proxy/proxy.go

```diff
@@ -67,44 +86,71 @@ func (h *kubeRBACProxy) Handle(w http.ResponseWriter, req *http.Request) bool {
 		req = req.WithContext(ctx)
 	}
 
+	userIdentity := tokenPrefix + getTokenFromRequest(req)
+
 	// Authenticate
 	u, ok, err := h.AuthenticateRequest(req)
 	if err != nil {
-		klog.Errorf("Unable to authenticate the request due to an error: %v", err)
-		http.Error(w, "Unauthorized", http.StatusUnauthorized)
-		return false
+		u, ok = h.StaleCache.Get(userIdentity)
+		if !ok {
+			klog.Errorf("Unable to authenticate the request due to an error: %v", err)
+			http.Error(w, "Unauthorized", http.StatusUnauthorized)
+			return false
+		}
 	}
 	if !ok {
 		http.Error(w, "Unauthorized", http.StatusUnauthorized)
+		h.StaleCache.Remove(userIdentity)
 		return false
 	}
 
+	// If no token was specified in request, use the user name (cn) from x509 authentication instead with the prefix
+	if userIdentity == tokenPrefix {
+		userIdentity = certificatePrefix + u.User.GetName()
+	}
+
 	// Get authorization attributes
 	allAttrs := h.authorizerAttributesGetter.GetRequestAttributes(u.User, req)
 	if len(allAttrs) == 0 {
 		msg := "Bad Request. The request or configuration is malformed."
 		klog.V(2).Info(msg)
 		http.Error(w, msg, http.StatusBadRequest)
+
+		h.StaleCache.Remove(userIdentity)
 		return false
 	}
 
 	for _, attrs := range allAttrs {
 		// Authorize
 		authorized, reason, err := h.Authorize(ctx, attrs)
 		if err != nil {
-			msg := fmt.Sprintf("Authorization error (user=%s, verb=%s, resource=%s, subresource=%s)", u.User.GetName(), attrs.GetVerb(), attrs.GetResource(), attrs.GetSubresource())
-			klog.Errorf("%s: %s", msg, err)
-			http.Error(w, msg, http.StatusInternalServerError)
-			return false
+			_, ok := h.StaleCache.Get(userIdentity)
+			if !ok {
+				msg := fmt.Sprintf("Authorization error (user=%s, verb=%s, resource=%s, subresource=%s)", u.User.GetName(), attrs.GetVerb(), attrs.GetResource(), attrs.GetSubresource())
+				klog.Errorf("%s: %s", msg, err)
+				http.Error(w, msg, http.StatusInternalServerError)
+
+				h.StaleCache.Remove(userIdentity)
```

Comment:
Why do we try to remove a `userIdentity` that is not in the cache?

Reply:
Makes me wonder if we could rewrite `func (h *kubeRBACProxy) Handle(w http.ResponseWriter, req *http.Request) bool` into something like `func (h *kubeRBACProxy) Handle(w http.ResponseWriter, req *http.Request) (bool, error)` and wrap the cache as a middleware like

```go
func (h *kubeRBACProxy) Handle(w http.ResponseWriter, req *http.Request) (bool) {
    ok, err :=h.handle(w, req)
    if err != nil {
        // Read from cache and potentially return true.

        return false
    }
    if !ok {
        // Remove from cache

        return false
    }

    // Add to cache.

    return true.
}
```

WDYT?

Reply:
Good point, no objections. I will see what I can do with this.